architecture/evm/error_normalizer.go

```diff
@@ -257,66 +301,78 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
-			strings.Contains(msg, "param is required") ||
-			strings.Contains(msg, "Invalid Request") ||
-			strings.Contains(msg, "validation errors") ||
-			strings.Contains(msg, "invalid argument") ||
-			strings.Contains(msg, "invalid params") {
-			if strings.Contains(msg, "tx of type") {
-				// TODO For now we're returning a server-side exception for this error
-				// so that the request is retried on a different upstream which supports the requested TX type.
-				// This must be properly handled when "Upstream Features" is implemented which allows for feature-based routing.
-				return common.NewErrEndpointServerSideException(
-					common.NewErrJsonRpcExceptionInternal(
-						int(code),
-						common.JsonRpcErrorCallException,
-						err.Message,
-						nil,
-						details,
-					),
-					nil,
-				)
-			}
-			return common.NewErrEndpointClientSideException(
+		}
+
+		//----------------------------------------------------------------
+		// "Unsupported" errors
+		//----------------------------------------------------------------
+
+		// Note: do not move this check above "Not found" errors, as we want to
+		// avoid premature detection when message is only "not found" (e.g. from Tenderly)
+
+		if r.StatusCode == 415 || code == common.JsonRpcErrorUnsupportedException || // By HTTP status code or explicit JSON-RPC error code
+			code == -32004 || code == -32001 || // direct codes from upstream
+			strings.Contains(msg, "Unsupported method") ||
+			strings.Contains(msg, "not supported") ||
+			strings.Contains(msg, "method is not whitelisted") ||
+			strings.Contains(msg, "is not included in your current plan") {
+			return common.NewErrEndpointUnsupported(
 				common.NewErrJsonRpcExceptionInternal(
 					int(code),
-					common.JsonRpcErrorInvalidArgument,
+					common.JsonRpcErrorUnsupportedException,
 					err.Message,
 					nil,
 					details,
 				),
 			)
-		} else if r.StatusCode == 415 || code == -32004 || code == -32001 || code == common.JsonRpcErrorUnsupportedException {
-			return common.NewErrEndpointUnsupported(
+		}
+
+		//----------------------------------------------------------------
+		// "Invalid Argument / Params / Request" errors
+		//----------------------------------------------------------------
+
+		// Even though these errors have invalid argument or params, they are more about a lack of standard
+		// for certain args/params for certain methods, among different providers or clients. We will retry them.
+		//
+		// Examples:
+		if strings.Contains(msg, "tx of type") || // Should be retried toward upstreams that support this tx type
+			strings.Contains(msg, "invalid type: map, expected BlockNumber, 'latest', or 'earliest'") { // Envio not supporting { blockNumber: XXX, blockHash: YYY } for eth_getBlockReceipts
+			return common.NewErrEndpointClientSideException(
 				common.NewErrJsonRpcExceptionInternal(
 					int(code),
-					common.JsonRpcErrorUnsupportedException,
+					common.JsonRpcErrorInvalidArgument,
```

Comment:
Still common.JsonRpcErrorUnsupportedException is more accurate is there a reason to make it InvalidArgument?

Reply:
hmm I think for the first one (`tx of type`) the `unsupported` makes more sense while for the second one the `invalid argument`. maybe best to separate those here?

Reply:
Even for second is basically a tag that they don't support, so it should be fine imo